server/api/auth.ts

```diff
@@ -0,0 +1,44 @@
+import crypto from "crypto";
+import express from "express";
+
+import { ApiSessionReq, ApiSessionRes } from "../../common/types";
+import { Sessions, Users } from "../models";
+import { attachDoubleSubmit, generateToken } from "../security";
+
+const authApi = express.Router();
+
+authApi.post("/session", async (req, res) => {
+  const params: ApiSessionReq = req.body;
+
+  if (!params.doubleSubmit || params.doubleSubmit !== req.cookies["doubleSubmit"]) {
+    await attachDoubleSubmit(res); // reattach since the only reasonable explanation is that the double submit cookie has expired
+    return res.status(400).json({ message: "Invalid double submit." });
+  }
+
+  const user = Users.findByEmail(params.email);
+  if (!user || !Users.verifyPassword(params.password, user)) {
+    await new Promise((resolve) => setTimeout(resolve, crypto.randomInt(11, 111))); // mitigate timing attacks
+    return res.status(401).json({ message: "Invalid credentials." });
+  }
+
+  const sessionId = await generateToken();
+  const csrfId = await generateToken();
+
+  res.cookie("sessionId", sessionId, { httpOnly: true, maxAge: 1000 * 3600 * 24, sameSite: "lax" });
+  res.cookie("csrfId", csrfId, { httpOnly: false, maxAge: 1000 * 3600 * 24, sameSite: "strict" });
+  Sessions.add({ csrfId, sessionId, userEmail: user.email });
```

Comment:
do we need both `csrfId` and `doubleSubmit` cookies? Or these are some left-overs after refactoring?

Reply:
I thought it might be useful to have a `csrf` cookie attached to a session (as opposed to generated on every request), but on a second thought, it doesn't seem all that useful at the moment, removed it for now.